tests/lib/rules/no-unsafe-optional-chaining.js

```diff
@@ -0,0 +1,231 @@
+/**
+ * @fileoverview Tests for no-unsafe-optional-chaining rule.
+ * @author Yeon JuAn
+ */
+
+"use strict";
+
+const rule = require("../../../lib/rules/no-unsafe-optional-chaining");
+
+const { RuleTester } = require("../../../lib/rule-tester");
+
+const parserOptions = {
+    ecmaVersion: 2021,
+    sourceType: "module"
+};
+
+const ruleTester = new RuleTester({ parserOptions });
+
+ruleTester.run("no-unsafe-optional-chaining", rule, {
+    valid: [
+        "var foo;",
+        "class Foo {}",
+        "!!obj?.foo",
+        "obj?.foo();",
+        "obj?.foo?.();",
+        "(obj?.foo ?? bar)();",
+        "(obj?.foo)?.()",
+        "(obj?.foo ?? bar?.baz)?.()",
+        "(obj.foo)?.();",
+        "obj?.foo.bar;",
+        "obj?.foo?.bar;",
+        "(obj?.foo)?.bar;",
+        "(obj?.foo)?.bar.baz;",
+        "(obj?.foo)?.().bar",
+        "(obj?.foo ?? bar).baz;",
+        "(obj?.foo ?? val)`template`",
+        "new (obj?.foo ?? val)()",
+        "new bar();",
+        "obj?.foo?.()();",
+        "const {foo} = obj?.baz || {};",
+        "bar(...obj?.foo ?? []);",
+        "var bar = {...foo?.bar};",
+        "foo?.bar in {};",
+        "[foo = obj?.bar] = [];",
+        "(foo?.bar, bar)();",
+        "(foo?.bar ? baz : qux)();",
+
+        // logical operations
+        "(obj?.foo ?? bar?.baz ?? qux)();",
+        "((obj?.foo ?? bar?.baz) || qux)();",
+        "((obj?.foo || bar?.baz) || qux)();",
+        "((obj?.foo && bar?.baz) || qux)();",
+
+        // The default value option disallowArithmeticOperators is false
+        "obj?.foo - bar;",
+        "obj?.foo + bar;",
+        "obj?.foo * bar;",
+        "obj?.foo / bar;",
+        "obj?.foo % bar;",
+        "obj?.foo ** bar;",
+        "+obj?.foo;",
+        "-obj?.foo;",
+        "bar += obj?.foo;",
+        "bar -= obj?.foo;",
+        "bar %= obj?.foo;",
+        "bar **= obj?.foo;",
+        "bar *= obj?.boo",
+        "bar /= obj?.boo",
+
+        {
+            code: "(obj?.foo || baz) + bar;",
+            options: [{
+                disallowArithmeticOperators: true
+            }]
+        },
+        {
+            code: "(obj?.foo ?? baz) + bar;",
+            options: [{
+                disallowArithmeticOperators: true
+            }]
+        },
+        {
+            code: "bar += obj?.foo ?? val",
+            options: [{
+                disallowArithmeticOperators: true
+            }]
+        }
```

Comment:
Maybe a couple of tests to show that `disallowArithmeticOperators: true` doesn't apply to all operators:

```js
{
    code: "typeof obj?.foo;",
    options: [{
        disallowArithmeticOperators: true
    }]
},
{
    code: "obj?.foo >> obj?.bar;",
    options: [{
        disallowArithmeticOperators: true
    }]
},
{
    code: "bar |= obj?.foo;",
    options: [{
        disallowArithmeticOperators: true
    }]
}
```